pkg/kapp/diff/stripnamehashsuffix.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package diff
+
+import (
+	"strings"
+
+	ctlconf "github.com/vmware-tanzu/carvel-kapp/pkg/kapp/config"
+	ctlres "github.com/vmware-tanzu/carvel-kapp/pkg/kapp/resources"
+)
+
+type stripNameHashSuffixConfig struct {
+	ResourceMatcher ctlres.ResourceMatcher
+}
+
+func (d stripNameHashSuffixConfig) EnabledFor(res ctlres.Resource) (stripEnabled bool) {
+	return d.ResourceMatcher.Matches(res)
+}
+
+func newStripNameHashSuffixConfig(resourceMatchers [][]ctlres.ResourceMatcher) (result stripNameHashSuffixConfig) {
+	result = stripNameHashSuffixConfig{}
+
+	var allMatchers []ctlres.ResourceMatcher
+	for _, matchers := range resourceMatchers {
+		// configurations that do not specify any matchers (like default
+		// config) have nil, which would lead AndMatcher to never match.
+		// as all the configs AndMatchers are ANDed, too, this would prevent
+		// matchers from any configuration to match, which would strip the
+		// whole suffix-strip configuration of any usefulness; as such we
+		// exclude nil matchers. :)
+		if matchers != nil {
+			allMatchers = append(allMatchers, ctlres.AndMatcher{
+				Matchers: matchers,
+			})
+		}
+	}
+	// N.B. if no config specifies any matchers (the default), then
+	// allMatchers will stay uninitialized/nil and the AndMatcher will never
+	// match, effectively disabling suffix strip.
+	result.ResourceMatcher = ctlres.AndMatcher{Matchers: allMatchers}
```

Comment:
I am still trying to understand why this is required. Would you be able to share a couple of scenarios that you are trying to cover, maybe we can find an alternative to it.

Reply:
We already have a notMatcher present. Do you think that will be useful instead of `exclude`? 
So, this how the matchers currently work. 
By default, any matcher is used. So you can just have something like this to match ConfigMaps and Secrets.
```yaml
stripNameHashSuffix:
- resourceMatchers:
  - apiVersionKindMatcher: {apiVersion: v1, kind: ConfigMap}
  - apiVersionKindMatcher: {apiVersion: v1, kind: Secret}
```
If you want to exclude something, then you can use an andMatcher (with anyMatcher and notMatcher nested inside), something like [this](https://github.com/vmware-tanzu/carvel-kapp/blob/develop/pkg/kapp/config/default.go#L224-L235). So I think we can rely on the existing capabilities matchers.

Reply:
The challenge then is aggregating such nested "and: (any , not)" matchers across configurations.

Reply:
> across configurations

Ideally we would want to have one Config only (although we could add as many as we want). Do you have any specific scenario where we would need multiple Configs?

Reply:
I thought this over, no use-case springs to mind; Ii wanted to include it for completeness and the include/exclude was the latest pragamtic way I found to implement it.

I would then explicitly implement it so that only the suffix-config from last kapp-config is used :)